cpu/atxmega/periph/uart.c

```diff
@@ -0,0 +1,379 @@
+
+        if (locBsel >= 4095) {
+            continue;
+        }
+
+        /* Omit division by 16 get higher accuracy at small baudrates*/
+        calcbaud = ((*fper) >> locBscale) / ((locBsel + 1) << (4 - clk2x));
+
+        if (_check_bsel( baud, &calcbaud, &precision)) {
+            *bsel = locBsel;
+            *bscale = locBscale;
+        }
+    }
+
+    /* More math for the negative equation
+     * bscale is negative so 1/2^bscale = 2^|bscale| which is a factor and
+     * not a division again runding the result before division with
+     * 0.5 * denominator
+     *
+     * bsel = 1/2^bscale *( fcpu / ( (16*baud)-1) )
+     *      = (  fcpu*2^|bscale| - (16*baud)*2^|bscale| )/(16*baud)
+     *      = (  fcpu*2^|bscale| - (16*baud)*2^|bscale| + 0.5*(16*baud) )/(16*baud)
+     *
+     * deno  =  (16/2*baud) = (baud<<(3-clk2x))
+     *
+     * bsel = (  (fcpu - (deno<<(1))<<|bscale|) + deno )/(deno<<1)
+     *
+     * Baud = (fper*1/2^bscale)/ (16*(bsel+1/2^bscale))
+     */
+    for (locBscale = -1; locBscale >= -7; locBscale--) {
+        uint32_t num = 0;
+
+        deno = ((*baud) << (3 - clk2x));
+        num = (((*fper) - (deno << 1)) << (-locBscale)) + deno;
+        num = (num / (deno << 1));
+
+        if (num >= 4095) {
+            break;
+        }
+
+        locBsel = (uint16_t)(num & 0xFFF);
+
+        /* Omit division by 16 get higher accuracy at small baudrates */
+        calcbaud = ((*fper) << (-locBscale)) /
+                   ((locBsel + (1 << (-locBscale))) << (4 - clk2x));
+
+        if (_check_bsel( baud, &calcbaud, &precision)) {
+            *bsel = locBsel;
+            *bscale = locBscale;
+        }
+    }
+
+    return precision;
+}
+
+/**
+ * @brief   Calculates bsel and bscale for a given periphery clock and baudrate.
+ *          Limitation are the periphery clock maximum is 32MHz, unsigned int
+ *          overflows if clock is bigger. And the periphery clock has to be not
+ *          smaller then 1 when divided by 128.
+ *
+ * fper/128 !=0 must be divide able by 128
+ * fper*128 != uint23_max => 32MHz max fper
+ */
+static inline int16_t _xmega_calculate_bsel_bscale(uint32_t fcpu, uint32_t baud,
+                                                   uint8_t *clk2x,
+                                                   uint16_t *bsel,
+                                                   int8_t *bscale)
+{
+    int16_t precision = 0;
+
+    precision = _xmega_bsel_bscale(&fcpu, &baud, 0, bsel, bscale );
+
+    /* default 2% precision, required precision is at least 2% */
+    if (precision <= (10000 + BAUD_TOL)) {
+        return (precision - 10000);
+    }
+
+    /* Precision goal was not met calculate baudrate with uart frequency doubled */
+    precision = _xmega_bsel_bscale(&fcpu, &baud, 1, bsel, bscale );
+    *clk2x = 1;
+
+    return (precision - 10000);
+}
+
+static inline void _configure_pins(uart_t uart)
+{
+    /* configure RX pin */
+    if (uart_config[uart].rx_pin != GPIO_UNDEF) {
```

Comment:
```suggestion
    if (gpio_is_valid(uart_config[uart].rx_pin)) {
```

in preparation for a new GPIO API